components/mediation-monitor/message-relay/org.wso2.carbon.relay/src/main/java/org/wso2/carbon/relay/ExpandingMessageFormatter.java

```diff
@@ -182,18 +182,18 @@ private void findAndWrite2OutputStream(MessageContext messageContext,
                 DataHandler dh = (DataHandler) binaryDataNode.getDataHandler();
 
                 DataSource dataSource = dh.getDataSource();
-                if (((StreamingOnRequestDataSource) dataSource).isConsumed()) {
-                    Object httpMethodObj = messageContext.getProperty(Constants.Configuration.HTTP_METHOD);
-                    if ((httpMethodObj instanceof String)
-                        && "POST".equals(httpMethodObj)) {
-                        log.warn("Attempting to send an already consumed request ["
-                                 + messageContext.getTo().getAddress()
-                                 + " POST/Empty Message Body]");
+                if (dataSource instanceof StreamingOnRequestDataSource) {
+                    if (((StreamingOnRequestDataSource) dataSource).isConsumed()) {
+                        Object httpMethodObj = messageContext.getProperty(Constants.Configuration.HTTP_METHOD);
+                        if ((httpMethodObj instanceof String) && "POST".equals(httpMethodObj)) {
+                            log.warn("Attempting to send an already consumed request [" +
+                                     messageContext.getTo().getAddress() + " POST/Empty Message Body]");
+                        }
+                    }
+                    //Ask the data source to stream, if it has not already cached the request
+                    if (!preserve && dataSource instanceof StreamingOnRequestDataSource) {
```

Comment:
"dataSource instanceof StreamingOnRequestDataSource" is redundant since it is already checked above in line 185.